--- a/lib/python3.10/site-packages/matplotlib/__init__.py
+++ b/lib/python3.10/site-packages/matplotlib/__init__.py
@@ -463,31 +463,18 @@ def _get_config_or_cache_dir(xdg_base_getter):
     configdir = os.environ.get('MPLCONFIGDIR')
     if configdir:
         configdir = Path(configdir).resolve()
-    elif sys.platform.startswith(('linux', 'freebsd')):
-        # Only call _xdg_base_getter here so that MPLCONFIGDIR is tried first,
-        # as _xdg_base_getter can throw.
-        configdir = Path(xdg_base_getter(), "matplotlib")
-    else:
-        configdir = Path.home() / ".matplotlib"
-    try:
-        configdir.mkdir(parents=True, exist_ok=True)
-    except OSError:
-        pass
-    else:
-        if os.access(str(configdir), os.W_OK) and configdir.is_dir():
-            return str(configdir)
+        try:
+            configdir.mkdir(parents=True, exist_ok=True)
+        except OSError:
+            pass
+        else:
+            if os.access(str(configdir), os.W_OK) and configdir.is_dir():
+                return str(configdir)
     # If the config or cache directory cannot be created or is not a writable
     # directory, create a temporary one.
     tmpdir = os.environ["MPLCONFIGDIR"] = \
         tempfile.mkdtemp(prefix="matplotlib-")
     atexit.register(shutil.rmtree, tmpdir)
-    _log.warning(
-        "Matplotlib created a temporary config/cache directory at %s because "
-        "the default path (%s) is not a writable directory; it is highly "
-        "recommended to set the MPLCONFIGDIR environment variable to a "
-        "writable directory, in particular to speed up the import of "
-        "Matplotlib and to better support multiprocessing.",
-        tmpdir, configdir)
     return tmpdir
 
 
-- 
2.26.1

